reporter/internal/pdata/generate.go

```diff
@@ -0,0 +1,322 @@
+	fileIDtoMapping := make(map[libpf.FileID]int32)
+
+	var locationIndex int32
+	var startTS, endTS pcommon.Timestamp
+	for traceKey, traceInfo := range events {
+		sample := profile.Sample().AppendEmpty()
+		sample.SetLocationsStartIndex(locationIndex)
+
+		slices.Sort(traceInfo.Timestamps)
+		startTS = pcommon.Timestamp(traceInfo.Timestamps[0])
+		endTS = pcommon.Timestamp(traceInfo.Timestamps[len(traceInfo.Timestamps)-1])
+
+		sample.TimestampsUnixNano().FromRaw(traceInfo.Timestamps)
+		sample.Value().Append(1)
+
+		// Walk every frame of the trace.
+		for i := range traceInfo.FrameTypes {
+			frameAttributes := addProfileAttributes(profile, []samples.AttrKeyValue[string]{
+				{Key: "profile.frame.type", Value: traceInfo.FrameTypes[i].String()},
+			}, attributeMap)
+
+			loc := profile.LocationTable().AppendEmpty()
+			loc.SetAddress(uint64(traceInfo.Linenos[i]))
+			loc.AttributeIndices().FromRaw(frameAttributes)
+
+			switch frameKind := traceInfo.FrameTypes[i]; frameKind {
+			case libpf.NativeFrame:
+				// As native frames are resolved in the backend, we use Mapping to
+				// report these frames.
+
+				var locationMappingIndex int32
+				if tmpMappingIndex, exists := fileIDtoMapping[traceInfo.Files[i]]; exists {
+					locationMappingIndex = tmpMappingIndex
+				} else {
+					idx := int32(len(fileIDtoMapping))
+					fileIDtoMapping[traceInfo.Files[i]] = idx
+					locationMappingIndex = idx
+
+					ei, exists := p.Executables.Get(traceInfo.Files[i])
+
+					// Next step: Select a proper default value,
+					// if the name of the executable is not known yet.
+					var fileName = "UNKNOWN"
+					if exists {
+						fileName = ei.FileName
+					}
+
+					mappingAttributes := addProfileAttributes(
+						profile,
+						[]samples.AttrKeyValue[string]{
+							// Once SemConv and its Go package is released with the new
+							// semantic convention for build_id, replace these hard coded
+							// strings.
+							{Key: "process.executable.build_id.gnu", Value: ei.GnuBuildID},
+							{Key: "process.executable.build_id.htlhash",
+								Value: traceInfo.Files[i].StringNoQuotes()},
+						},
+						attributeMap,
+					)
+
+					mapping := profile.MappingTable().AppendEmpty()
+					mapping.SetMemoryStart(uint64(traceInfo.MappingStarts[i]))
+					mapping.SetMemoryLimit(uint64(traceInfo.MappingEnds[i]))
+					mapping.SetFileOffset(traceInfo.MappingFileOffsets[i])
+					mapping.SetFilenameStrindex(getStringMapIndex(stringMap, fileName))
+					mapping.AttributeIndices().FromRaw(mappingAttributes)
+				}
+				loc.SetMappingIndex(locationMappingIndex)
+			case libpf.AbortFrame:
+				// Next step: Figure out how the OTLP protocol
+				// could handle artificial frames, like AbortFrame,
+				// that are not originated from a native or interpreted
+				// program.
+			default:
+				// Store interpreted frame information as a Line message:
+				line := loc.Line().AppendEmpty()
+
+				fileIDInfoLock, exists := p.Frames.Get(traceInfo.Files[i])
```

Comment:
Fix from https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/256:
```suggestion
				fileIDInfoLock, exists := p.Frames.GetAndRefresh(traceInfo.Files[i], ...)
```